jsonrpc/server.go

```diff
@@ -316,37 +335,52 @@ func (s *Server) handleRequest(req *request) (*response, error) {
 	return res, nil
 }
 
-func (s *Server) buildArguments(params, handler any, configuredParams []Parameter) ([]reflect.Value, error) {
-	args := make([]reflect.Value, 0, len(configuredParams))
+func (s *Server) buildArguments(ctx context.Context, params any, method Method) ([]reflect.Value, error) {
+	handlerType := reflect.TypeOf(method.Handler)
+	numArgs := handlerType.NumIn()
+
+	args := make([]reflect.Value, numArgs)
 	if isNil(params) {
-		if len(configuredParams) > 0 {
+		if len(method.Params) > 0 {
 			return nil, errors.New("missing non-optional param field")
 		}
 
 		return args, nil
 	}
 
-	handlerType := reflect.TypeOf(handler)
+	numNonContextArgs := numArgs
+	if method.needsContext {
+		numNonContextArgs--
+		// Insert context at beginning of args slice.
+		ctxV := reflect.ValueOf(ctx)
+		args[0] = ctxV
+	}
 
 	switch reflect.TypeOf(params).Kind() {
 	case reflect.Slice:
 		paramsList := params.([]any)
 
-		if len(paramsList) != handlerType.NumIn() {
+		if len(paramsList) != numNonContextArgs {
 			return nil, errors.New("missing/unexpected params in list")
 		}
 
 		for i, param := range paramsList {
+			if method.needsContext {
+				i++
+			}
```

Comment:
If you keep the previous `append` approach, this becomes unnecessary. You can just append the context to the args first, then continue appending `paramsList-1` number of actual arguments.

Reply:
since the entire list is pre-allocated `args := make([]reflect.Value, 0, len(configuredParams))` appending doesnt cause any allocations